main/src/ca/uwaterloo/flix/language/phase/Optimizer.scala

```diff
@@ -41,14 +40,255 @@ object Optimizer extends Phase[SimplifiedAst.Root, SimplifiedAst.Root] {
-      println(PrettyPrinter.Simplified.fmtRoot(root).fmt(TerminalContext.AnsiTerminal))
+
+    /**
+      * Returns `true` if and only if `enum` is a singleton-valued enum.
+      */
+    def isSingleCaseEnum(enum: Symbol.EnumSym): Boolean = root.enums.get(enum) match {
+      case None => false
+      case Some(defn) => defn.cases.size <= 1
     }
 
-    // TODO: Implement.
+    def toExp(b0: Boolean): Expression = if (b0) Expression.True else Expression.False
 
-    root.toSuccess
-  }
+    def optimizeBinaryExpression(e0: Expression.Binary): Expression = e0 match {
+      case Expression.Binary(op, exp1, exp2, tpe, loc) =>
+        val e1 = optimize(exp1)
+        val e2 = optimize(exp2)
+        (op, e1, e2) match {
+          // Elimination of run-time tag checks of singleton-valued enums.
+          case (BinaryOperator.Equal, Expression.Tag(sym, tag, exp, tp, lc), _) =>
+            if (isSingleCaseEnum(sym) && exp == Expression.Unit) Expression.True
+            else Expression.Binary(op, e1, e2, tpe, loc)
+          case (BinaryOperator.Equal, _, Expression.Tag(sym, tag, exp, tp, lc)) =>
+            if (isSingleCaseEnum(sym) && exp == Expression.Unit) Expression.True
+            else Expression.Binary(op, e1, e2, tpe, loc)
+          // Optimizations for Bool.
+          case (BinaryOperator.Equal | BinaryOperator.LogicalAnd, Expression.True, exp) => exp
+          case (BinaryOperator.Equal | BinaryOperator.LogicalAnd, exp, Expression.True) => exp
+          case (BinaryOperator.Equal, Expression.False, exp) => Expression.Unary(UnaryOperator.LogicalNot, exp, tpe, loc)
+          case (BinaryOperator.Equal, exp, Expression.False) => Expression.Unary(UnaryOperator.LogicalNot, exp, tpe, loc)
+          case (BinaryOperator.NotEqual, Expression.True, exp) => Expression.Unary(UnaryOperator.LogicalNot, exp, tpe, loc)
+          case (BinaryOperator.NotEqual, exp, Expression.True) => Expression.Unary(UnaryOperator.LogicalNot, exp, tpe, loc)
+          case (BinaryOperator.NotEqual | BinaryOperator.LogicalOr, Expression.False, exp) => exp
+          case (BinaryOperator.NotEqual | BinaryOperator.LogicalOr, exp, Expression.False) => exp
+          case (BinaryOperator.LogicalAnd, Expression.False, _) | (BinaryOperator.LogicalAnd, _, Expression.False) => Expression.False
+          case (BinaryOperator.LogicalOr, _, Expression.True) | (BinaryOperator.LogicalOr, Expression.True, _) => Expression.True
+          // Optimizations for Char.
+          case (_, Expression.Char(lit1), Expression.Char(lit2)) =>
+            op match {
+              case BinaryOperator.Equal => toExp(lit1 == lit2)
+              case BinaryOperator.NotEqual => toExp(lit1 != lit2)
+              case _ => Expression.Binary(op, e1, e2, tpe, loc)
+            }
+          // Optimizations for Float32.
+          case (_, Expression.Float32(lit1), Expression.Float32(lit2)) =>
+            op match {
+              case BinaryOperator.Equal => toExp(lit1 == lit2)
```

Comment:
Equality is okay for primitive types, its only for enums (tuples etc) I don't want to mess with it.